dkron/store.go

```diff
@@ -627,27 +545,38 @@ func (s *Store) SetExecution(execution *Execution) (string, error) {
 }
 
 // DeleteExecutions removes all executions of a job
-func (s *Store) DeleteExecutions(jobName string) error {
-	prefix := fmt.Sprintf("executions/%s/", jobName)
-	return s.db.DropPrefix([]byte(prefix))
+func (s *Store) deleteExecutionsTxFunc(jobName string) func(tx *buntdb.Tx) error {
+	return func(tx *buntdb.Tx) error {
+		var delkeys []string
+		prefix := fmt.Sprintf("executions/%s", jobName)
+		tx.Ascend("", func(key, value string) bool {
+			if strings.HasPrefix(key, prefix) {
+				delkeys = append(delkeys, key)
+			}
+			return true
+		})
+
+		for _, k := range delkeys {
+			_, _ = tx.Delete(k)
+		}
+
+		return nil
+	}
 }
 
 // Shutdown close the KV store
 func (s *Store) Shutdown() error {
 	return s.db.Close()
 }
 
-// Snapshot creates a backup of the data stored in Badger
+// Snapshot stub for Raft
 func (s *Store) Snapshot(w io.WriteCloser) error {
-	_, err := s.db.Backup(w, 0)
-	return err
+	return nil
 }
 
-// Restore load data created with backup in to Badger
-// Default value for maxPendingWrites is 256, to minimise memory usage
-// and overall finish time.
+// Restore stub for Raft
 func (s *Store) Restore(r io.ReadCloser) error {
```

Comment:
You can also remove this as we don't use this as Raft store

Reply:
If I remove this method (same for Snapshot), then other things claim about Restore/Snapshot methods.

```
diff --git a/dkron/fsm.go b/dkron/fsm.go
index 03474bf3..69505961 100644
--- a/dkron/fsm.go
+++ b/dkron/fsm.go
@@ -139,7 +139,7 @@ func (d *dkronFSM) Snapshot() (raft.FSMSnapshot, error) {
 // Restore stores the key-value store to a previous state.
 func (d *dkronFSM) Restore(r io.ReadCloser) error {
        defer r.Close()
-       return d.store.Restore(r)
+       return nil
 }
 
 type dkronSnapshot struct {
@@ -147,12 +147,6 @@ type dkronSnapshot struct {
 }
 
 func (d *dkronSnapshot) Persist(sink raft.SnapshotSink) error {
-       if err := d.store.Snapshot(sink); err != nil {
-               sink.Cancel()
-               return err
-       }
-
-       // Close the sink.
        if err := sink.Close(); err != nil {
                return err
        }
```

Should I return nil in place like above instead?
Because, if I remove Snapshot and Persist in fsm.go, then raft.NewRaft() and raft.RecoverCluster() claim that they're missing methods Restore and Snapshot:
```
cannot use fsm (variable of type *dkronFSM) as raft.FSM value in argument to raft.NewRaft: missing method Restore
cannot use tmpFsm (variable of type *dkronFSM) as raft.FSM value in argument to raft.RecoverCluster: missing method Restore
```

Reply:
Oh yes, because it's defined in the Storage interface which FSM uses. Restore and Snapshot should be implemented with https://godoc.org/github.com/tidwall/buntdb#DB.Load and https://godoc.org/github.com/tidwall/buntdb#DB.Save

Reply:
I wonder what's the point having Save and Load if BuntDB is restored through raft which is persisted and even replicated from other nodes?

Reply:
This is in case of server restart, the point is that Raft log works with a combination of snapshots and log, when you restart a server that has been compacted, it will load the snapshot and apply the logs since the last snapshot. If snapshots are not saved, the storage won't be in sync.

Reply:
Alright, gotcha. My confusion was caused, because we run Dkron in Kubernetes without any state persistence. Each pod restart causes full wipe both of Raft and Store.
We just rely on HA and replication there. In case everything is really bad, we just restore job definitions with Terraform.